plugin/trino-hive/src/main/java/io/trino/plugin/hive/avro/TrinoAvroSerDe.java

```diff
@@ -13,29 +13,177 @@
+                throw new AvroSerdeException(EXCEPTION_MESSAGE);
+            }
+            final String columnNameDelimiter = properties.containsKey(serdeConstants.COLUMN_NAME_DELIMITER) ? properties
+                    .getProperty(serdeConstants.COLUMN_NAME_DELIMITER) : String.valueOf(SerDeUtils.COMMA);
+            // Get column names and types
+            List<String> columnNames = Arrays.asList(columnNameProperty.split(columnNameDelimiter));
+            List<TypeInfo> columnTypes = TypeInfoUtils.getTypeInfosFromTypeString(columnTypeProperty);
+
+            Schema schema = AvroSerDe.getSchemaFromCols(properties, columnNames, columnTypes, columnCommentProperty);
+            properties.setProperty(AvroTableProperties.SCHEMA_LITERAL.getPropName(), schema.toString());
+            if (conf != null) {
+                conf.set(AvroTableProperties.AVRO_SERDE_SCHEMA.getPropName(), schema.toString(false));
+            }
+            return schema;
+        }
+        else if (schemaString.equals(SCHEMA_NONE)) {
+            throw new AvroSerdeException(EXCEPTION_MESSAGE);
+        }
+
+        try {
+            Schema s = getSchemaFromFS(schemaString, conf);
+            if (s == null) {
+                //in case schema is not a file system
+                return AvroSerdeUtils.getSchemaFor(new URL(schemaString));
+            }
+            return s;
+        }
+        catch (IOException ioe) {
+            throw new AvroSerdeException("Unable to read schema from given path: " + schemaString, ioe);
+        }
+        catch (URISyntaxException urie) {
+            throw new AvroSerdeException("Unable to read schema from given path: " + schemaString, urie);
+        }
+    }
+
+    /**
+     * Copied from {@code AvroSerdeUtils}. No changes were made except formatting and
+     * calling a method from this class instead of that one.
+     */
+    private static Schema getSchemaFromFS(String schemaFSUrl, Configuration conf)
+            throws IOException, URISyntaxException
+    {
+        FSDataInputStream in = null;
+        FileSystem fs = null;
+        try {
+            fs = FileSystem.get(new URI(schemaFSUrl), conf);
+        }
+        catch (IOException ioe) {
+            //return null only if the file system in schema is not recognized
+            LOG.debug(ioe, "Failed to open file system for uri %s assuming it is not a FileSystem url", schemaFSUrl);
+            return null;
+        }
+        try {
+            in = fs.open(new Path(schemaFSUrl));
+            Schema s = getSchemaFor(in); // This line changed from AvroSerdeUtils
+            return s;
+        }
+        finally {
+            if (in != null) {
+                in.close();
+            }
+        }
+    }
+
+    /**
+     * Copied from {@link AvroSerdeUtils#getSchemaFor(InputStream) AvroSerdeUtils}.
+     * The only non-formatting change is the addition of {@code .setValidateDefaults(false)}.
```

Comment:
... then the only non-formatting change will be visible as a code change, in normal fix commit.

However, is it possible to avoid copying so much code from Hive?
Can we e.g. propose  a change in apache/hive which would allow us to remove the code copy?

Reply:
>  Can we e.g. propose a change in apache/hive which would allow us to remove the code copy?

This does seem like a good idea; would we just want to open a new issue on Hive? Or maybe comment on the [HIVE-24796](https://issues.apache.org/jira/browse/HIVE-24797) asking why it wasn't added to Hive 3?

Reply:
cont'd at https://github.com/trinodb/trino/pull/9243#issuecomment-925699388